src/hazelcore/semantics/HTyp.re

```diff
@@ -1,69 +1,55 @@
+open Sexplib.Std;
+
 /* types with holes */
 [@deriving sexp]
 type t =
   | Hole
-  | Unit
   | Num
   | Bool
   | Arrow(t, t)
-  | Prod(t, t)
   | Sum(t, t)
+  | Prod(list(t))
   | List(t);
 
-// Prod(Num, Num)  ==  (Num, Num)
-// (Prod(Num, Num), Bool)  ==  (Num, Num, Bool)
-
-let rec num_tms = (ty: t): int =>
+let precedence_const = 0;
+let precedence_Prod = 1;
+let precedence_Sum = 2;
+let precedence_Arrow = 3;
+let precedence = (ty: t): int =>
   switch (ty) {
-  | Hole
-  | Unit
   | Num
   | Bool
-  | List(_) => 1
-  | Arrow(ty1, ty2)
-  | Prod(ty1, ty2)
-  | Sum(ty1, ty2) => num_tms(ty1) + num_tms(ty2)
+  | Hole
+  | Prod([])
+  | List(_) => precedence_const
+  | Prod(_) => precedence_Prod
+  | Sum(_, _) => precedence_Sum
+  | Arrow(_, _) => precedence_Arrow
   };
 
-/* eqity */
-let rec eq = (ty1, ty2) =>
-  switch (ty1, ty2) {
-  | (Hole, Hole) => true
-  | (Hole, _) => false
-  | (Unit, Unit) => true
-  | (Unit, _) => false
-  | (Num, Num) => true
-  | (Num, _) => false
-  | (Bool, Bool) => true
-  | (Bool, _) => false
-  | (Arrow(ty1, ty2), Arrow(ty1', ty2')) => eq(ty1, ty1') && eq(ty2, ty2')
-  | (Arrow(_, _), _) => false
-  | (Prod(ty1, ty2), Prod(ty1', ty2')) => eq(ty1, ty1') && eq(ty2, ty2')
-  | (Prod(_, _), _) => false
-  | (Sum(ty1, ty2), Sum(ty1', ty2')) => eq(ty1, ty1') && eq(ty2, ty2')
-  | (Sum(_, _), _) => false
-  | (List(ty), List(ty')) => eq(ty, ty')
-  | (List(_), _) => false
-  };
+/* eqity
+   At the moment, this coincides with default equality,
+   but this will change when polymorphic types are implemented */
+let eq = (==);
 
 /* type consistency */
 let rec consistent = (x, y) =>
   switch (x, y) {
   | (Hole, _)
   | (_, Hole) => true
-  | (Unit, Unit) => true
-  | (Unit, _) => false
   | (Num, Num) => true
   | (Num, _) => false
   | (Bool, Bool) => true
   | (Bool, _) => false
   | (Arrow(ty1, ty2), Arrow(ty1', ty2'))
-  | (Prod(ty1, ty2), Prod(ty1', ty2'))
   | (Sum(ty1, ty2), Sum(ty1', ty2')) =>
     consistent(ty1, ty1') && consistent(ty2, ty2')
   | (Arrow(_, _), _) => false
-  | (Prod(_, _), _) => false
   | (Sum(_, _), _) => false
+  | (Prod(tys1), Prod(tys2)) =>
+    ListUtil.for_all2_opt(consistent, tys1, tys2)
```

Comment:
Like this you can stil differentiate between the cases where the lists have different lengths and where the predicate is false. I think if the default value when the lists are of different lengths would always be `false` it would not matter, but I think the default might be `true` as well sometimes.